### ISSUE TITLE
Add HashedTimerWheel implementation

### ISSUE THREAD

**Issue by sirthias** [JCTools/JCTools#109]
e.g. something like Agrona's:
https://github.com/real-logic/Agrona/blob/master/src/main/java/org/agrona/TimerWheel.java

While not strictly a concurrent data structure itself it does appear to be frequently re-implemented by what would be the target user base of JCTools, e.g. [Netty](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/HashedWheelTimer.java), [Agrona](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/TimerWheel.java), [Reactor](https://github.com/reactor/reactor-core/blob/master/src/main/java/reactor/core/timer/HashWheelTimer.java) or [Kafka](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/utils/timer/Timer.scala).

Having a solid default implementation should help everyone focus on their real goals.

**Comment by nitsanw** [comment]
pinging @mjpt777 
There's some overlap between Agrona and JCTools, I would like to have Agrona depend on JCTools for queues and such but have been thus far unsuccessful in convincing Martin. Bringing over HashedTimerWheel will increase the duplication which I'm not too thrilled about. On the other hand, if it makes sense to include it then maybe we should.
I would value some feedback from relevant project maintainers: @mjpt777 @normanmaurer @jrudolph @hachikuji @smaldini

**Comment by RichardWarburton** [comment]
Can you explain why you don't use Agrona's timer wheel? Is it wanting to avoid an additional dependency for a single class?

**Comment by mjpt777** [comment]
I'm happy to donate the TimerWheel code from Agrona across to JCTools if that makes most sense. In Aeron we found it was better to not have the indirection in the end.

**Comment by sirthias** [comment]
@RichardWarburton Yes, I'd like to avoid an extra dependency (in my case each dependency is a careful trade-off).
Also, JCTools appears to me more as a "primary" kind of library which isn't a spin-off of some parent project and therefore more likely to receive (and accept) contributions that don't have an immediate benefit to the parent project.
(This is not to say that Agrona isn't a valuable library by itself!
I'd simply prefer to depend on JCTools instead in all cases where there's overlap.)

**Comment by nitsanw** [comment]
@sirthias I've had a read through the code and will have to think about this. The scope for JCTools at the moment is for concurrent data structures, and this is somewhere on the boundary.

**Comment by sirthias** [comment]
@nitsanw No problem. Thanks for considering this feature request, independently of the inclusion decision!

**Comment by ifesdjeen** [comment]
I've worked on the Reactor hashed wheel implementation, and have recently extracted it to the standalone library: https://github.com/ifesdjeen/hashed-wheel-timer 

Thought it might be relevant. I've looked at Netty, Reactor, Agrona and Kafka implementations and tried multiple variations in terms of simplicity and performance. I also included some jmh benchmarks.

**Comment by guidomedina** [comment]
For what I could read on your project description `MpscChunkedArrayQueue` could be ideal to rule them all, if GC is a concern.

**Comment by nitsanw** [comment]
@ifesdjeen @sirthias @guidomedina I've had a look at the separate repo. The timer implementation is significantly different, but it did help me in better setting the context for the requirements, so thanks.
I can see how a marriage of the Agrona timer and an MpscChunkedArrayQueue variant will result in a nice result. I would release it as a separate artifact. Maybe the right approach here is to mini-shade Agrona+JCTools to construct the scheduled executor in an all in one artifact?
I realize this is still an extra dep for @sirthias, but it would be a single dependency hiding the underlying grabbing and stitching of resources.

**Comment by sirthias** [comment]
@nitsanw Interesting. I am currently using an internalized copy of the Agrona implementation and realized that the timerwheel implementation really isn't all that much logic / complexity at all (especially after I've stripped all the bits that I don't need).
Do you think it really makes sense to package it as a separate artifact? A single (and rather small) file addition to JCTools would be all that's needed, IMHO.

**Comment by guidomedina** [comment]
@nitsanw I believe that it would be a good fit for JCtools; for a couple of reasons:
- The queue that it will use is part of JCtools.
- Most projects that use timers are by nature concurrent.
- People will understand all the juice that JCtools has in it once they realize they have one of the best non-blocking queues available and soon non-blocking hash maps with primitive specialization.

Look at Netty for example, they are already (in-progress) shading JCtools, they will find out that they can get rid of their timer, I don't know, maybe I'm like JCtools too much and my opinion is mostly bias ;-)

**Comment by nitsanw** [comment]
@sirthias are you expecting to use it as a scheduled executor replacement? I have plans for some executor replacements on top of JCTools (@mjpt777 was also keen to contribute at one point, but paying customers got in the way), but I always thought to release those as a split artifact (and maybe that's not a good idea?).

**Comment by sirthias** [comment]
@nitsanw Yes, I am using it in that way. Like everyone else my use case is efficient management of many 1000s of timeouts, where I can trade-off timer accuracy against resource consumption.

Ideally JCTools would provide a an efficient Scheduler/Executor which lets me configure
- the timer granularity (tick-duration)
- the timer wheel size (ticks-per-wheel)
- the number of threads backing the executor (most people appear to be using only a single one, but being able to distribute across several would allow for another degree of configuration freedom)

Of course scheduled timeouts should be cancellable and the overhead (memory and CPU) of each additional scheduled timeout should be as small as possible.

Most implementations I've seen are basically Netty's HashedWheelTimer with small adaptations to the respective environment to flatten composition levels in order to get around unnecessary allocations.
IMHO the challenge will be to come up with an implementation that is good, flexible and open enough to not trigger people's urge to once again embed an adapted copy in their projects.

**Comment by chrisvest** [comment]
I also have a use case for something like this (an efficient low-GC scheduled executor replacement) in Stormpot, plus I'd like to be able to inject my own monotonic clock.

**Comment by nitsanw** [comment]
<grumble, grumble> OK, any volunteers for a SchedueledExcutor PR? put it in experimental for now? With any luck it will be in the up and coming 1.3 release (in a month or so)

**Comment by guidomedina** [comment]
Another copy & paste:
https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala#L23

**Comment by ben-manes** [comment]
I've been meaning to implement this as well, for quite a while, to support variable cache expiration (Caffeine). I currently used fixed expiration since that is O(1) by using access/write-order linked lists.

I wanted to port an O(1) [tickless](http://www.25thandclement.com/~william/projects/timeout.c.html) timer wheel to avoid some of the [rebalancing delays](https://lwn.net/Articles/646950/).

I haven't had the time to start on this, but its been rumbling in my head for quite a while. I wouldn't mind using JCTools if it gets to it first :-)

**Comment by nitsanw** [comment]
@ben-manes **OR**....**or**..._or_... you could implement it in JCTools? If you believe it's the right place for it, it would be greatly appreciated. I might even buy you a beer, or give you a hug, or both!

**Comment by sirthias** [comment]
@ben-manes The alternative to the hashed-wheel-timer proposed by Thomas Gleixner for the Linux kernel really does look like a great solution, thanks for the pointer!
Since its properties are somewhat different from the "traditional" (2005) wheel implementation I guess people would like the ability to pick the one that better fits their needs.

@nitsanw So now we already have two appealing ideas for scheduled executor replacements that JCTools could provide... :)

**Comment by ben-manes** [comment]
@sirthias Thomas Gleixner's do look good and I wonder how it has evolved. The tickless timer by William Ahern is also very attractive, as he claims it is O(1) in all operations. I was hoping to dig into his version more.

In particular I'd really like a version that does not require a dedicated background thread. In my case as expiration it could be lazily applied and performed in a maintenance phase (periodically scheduled on ForkJoinPool#commonPool() based on usage). That trick is how the cache supports max-size, expiration, and reference collection without any dedicated threads. In earlier versions (CLHM, Guava) it amortized on the calling thread. I'd like to do similar with the timer wheel, so that would be an attractive API option.

@nitsanw Sure, when I get to it then I'll contribute the code and collect on that hug.

**Comment by sirthias** [comment]
@ben-manes It sounds like your use case is "Deferrable Timers" in Linux kernel parlance, for which special support is certainly an interesting aspect of a prospective JCTools solution.
AFAIK at least Netty and Akka have no such facility and simply treat Deferrable Timers just like any other timeout, thereby foregoing the efficiency benefits that special support would be able to deliver.

**Comment by rkuhn** [comment]
Thomas Gleixner’s implementation indeed looks interesting; the described main use-case also exactly matches what Akka needs—we have been more or less happy with our own Netty rip-off :-) but a large part of that is lingering ignorance of the real performance numbers: how do you benchmark a timer, in particular when the most interesting case is that it does not fire?

@sirthias I don’t understand the use-case for deferrable timers in Akka: when we schedule something, we want it to happen soon after the designated time. If it weren’t time-dependent then we’d just throw it into a threadpool. The Linux kernel can do more interesting things because it controls the CPU—which in turn brings more responsibility—but we cannot even hope to do that in a library on top of the JVM; heck, we cannot even preempt a task :-)

**Comment by sirthias** [comment]
@rkuhn AFAIU the idea behind Deferrable Timers is the acknowledgement that not all timers are created equal. For some timers I'd like the timeout period to be hit relatively accurately (e.g. for triggering a tick source in Akka Stream) and sometimes I schedule a timeout merely to protect against a potential resource leak (e.g. in an Akka Actor "ask"). In the latter case I might set the timeout to 30 secs but don't really care whether it'll only be triggered after 40 sec or even a minute. I simply want to make sure that it *will* be triggered sometime in the future, when the trigger can be run with minimal resource overhead.
In these cases the actual timeout value specified kind of defines the magnitude or lower bound of the timeout period, rather than a precise value.
By lumping these kinds of timeouts together and deprioritizing them behind the more precise variant one should be able to increase the capacity of a single-thread scheduler.

> how do you benchmark a timer, in particular when the most interesting case is that it does not fire?

I guess you could do it by measuring the *capacity* of a particular scheduler implementation by steadily increasing the timeout load (for a given distribution of timeout periods, cancellation rate and, potentially, timeout type) and noting when things begin to topple over.

**Comment by ben-manes** [comment]
I caught up with Doug, so this is down to which of us can procrastinate the most!

```
... A wheel-like utility has been on the j.u.c todo list for years with
low priority. This summer I might finally get a chance to see if all
the reasons for not doing it before can be made to disappear.
```

**Comment by ben-manes** [comment]
I wrote a simple [hierarchical timer wheel](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/TimerWheel.java) to iterate on, so I imagine it will change quite a bit as I flush out variable cache expiration. It is very much coupled to my usage, e.g. deferred timers, using fields on the cache entry, operated on during the maintenance operations (single writer). But I think it shows the general idea of how a hierarchical wheel operates and uses the normal power-of-two tricks.

I'll probably look at the other implementations to borrow ideas when optimizing.
